lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -72,6 +72,56 @@ CrossOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
   return success();
 }
 
+LogicalResult FilterOp::verifyRegions() {
+  MLIRContext *context = getContext();
+  Type si1 = IntegerType::get(context, /*width=*/1, IntegerType::Signed);
+  Region &condition = getCondition();
+
+  // Verify that type of yielded value is Boolean.
+  auto yieldOp = llvm::cast<YieldOp>(condition.front().getTerminator());
+  Type yieldedType = yieldOp.getValue().getType();
+  if (yieldedType != si1)
+    return emitOpError()
+           << " must have 'condition' region yielding 'si1' (yields "
+           << yieldedType << ")";
+
+  // Verify that block has argument of input tuple type.
+  Type tupleType = getResult().getType();
+  if (condition.getNumArguments() != 1 ||
+      condition.getArgument(0).getType() != tupleType) {
+    InFlightDiagnostic diag = emitOpError()
+                              << "must have 'condition' region taking "
+                              << tupleType << " as argument (takes ";
+    if (condition.getNumArguments() == 0)
+      diag << "no arguments)";
+    else
+      diag << condition.getArgument(0).getType() << ")";
+    return diag;
+  }
+
+  return success();
+}
+
+LogicalResult
+LiteralOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
+                            ValueRange operands, DictionaryAttr attributes,
+                            OpaqueProperties properties, RegionRange regions,
+                            llvm::SmallVectorImpl<Type> &inferredReturnTypes) {
+  auto *typedProperties = properties.as<Properties *>();
+  if (!loc)
+    loc = UnknownLoc::get(context);
+
+  auto attr = llvm::dyn_cast<TypedAttr>(typedProperties->getValue());
+  if (!attr)
+    ::emitError(loc.value()) << "unsuited attribute for literal value: "
```

Comment:
Use emitOptionalError here (its technically a bit of a mishap that I need to fix, but the intention was that one can run inference without reporting errors by not specifying a location - I was going to remove it, but folks found it semi useful)

Reply:
Good pointer! Yeah, I *do* find it useful, too! Changed!

I'd prefer if it returned an `InflightDiagnostics`, though, such that you can `<<` further information into the result of the call. I did see that you can add other `Args...`, so I can still emit stuff, but I kinda prfer the `<<` version...

I also realized that the original code did not return the error it emitted! I fixed that and added a test...